File: python/trees.py
# -*- coding: utf-8 -*-


class Node(object):
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def __iter__(self):
        if self.left:
            yield from self.left
        yield self.value
        if self.right:
            yield from self.right

    def __len__(self):
        return sum(1 for _ in self)

    def __reversed__(self):
        if self.right:
            yield from reversed(self.right)
        yield self.value
        if self.left:
            yield from reversed(self.left)

    @property
    def height(self):
        return max(
            self.left.height if self.left else 0,
            self.right.height if self.right else 0
        ) + 1

    def find_min(self):
        return self.left.find_min() if self.left else self

    def find_max(self):
        return self.right.find_max() if self.right else self

    def find(self, value):
        if value < self.value:
            if self.left:
                return self.left.find(value)
        elif value > self.value:
            if self.right:
                return self.right.find(value)
        else:
            return self

    def delete(self, value):
        if value < self.value:
            if self.left:
                self.left = self.left.delete(value)
        elif value > self.value:
            if self.right:
                self.right = self.right.delete(value)
        else:
            if self.left and self.right:
                child = self.right.find_min()
                self.value = child.value
                self.right = self.right.delete(child.value)
            elif self.left:
                return self.left
            elif self.right:
                return self.right
            else:
                return None
        return self

    def insert(self, value):
        if value < self.value:
            if self.left:
                self.left.insert(value)
            else:
                self.left = Node(value)
        elif value > self.value:
            if self.right:
                self.right.insert(value)
            else:
                self.right = Node(value)

    def leaves(self):
        if self.left or self.right:
            if self.left:
                yield from self.left.leaves()
            if self.right:
                yield from self.right.leaves()
        else:
            yield self.value

    def paths(self):
        def mkpath(path):
            yield self.value
            yield from path
        if self.left or self.right:
            if self.left:
                for path in self.left.paths():
                    yield tuple(mkpath(path))
            if self.right:
                for path in self.right.paths():
                    yield tuple(mkpath(path))
        else:
            yield tuple((self.value,))

    def __str__(self):
        if self.left or self.right:
            return '({} {} {})'.format(
                self.value,
                self.left or 'nil',
                self.right or 'nil'
            )
        return str(self.value)


class Tree(object):
    def __init__(self):
        self.root = None

    def __iter__(self):
        if self.root:
            yield from self.root

    def __len__(self):
        return len(self.root) if self.root else 0

    def __reversed__(self):
        if self.root:
            yield from reversed(self.root)

    @property
    def height(self):
        return self.root.height if self.root else 0

    def find_min(self):
        if self.root:
            return self.root.find_min()

    def find_max(self):
        if self.root:
            return self.root.find_max()

    def find(self, value):
        if self.root:
            return self.root.find(value)

    def delete(self, value):
        if self.root:
            self.root = self.root.delete(value)
        else:
            self.root = None
        return self.root

    def insert(self, value):
        if self.root:
            self.root.insert(value)
        else:
            self.root = Node(value)
        return self.root

    def leaves(self):
        if self.root:
            return tuple(self.root.leaves())
        return tuple()

    def paths(self):
        if self.root:
            return tuple(self.root.paths())
        return tuple()

    def __str__(self):
        if self.root:
            return str(self.root)
        return '<empty>'

File: julia/Algorithms/src/trees.jl
mutable struct BinaryNode
    value
    left::Union{BinaryNode, Nothing}
    right::Union{BinaryNode, Nothing}

    BinaryNode(value) = new(value, nothing, nothing)
end

mutable struct BinaryTree
    root::Union{BinaryNode, Nothing}

    BinaryTree() = new(nothing)

    function BinaryTree(values...)
        root = nothing
        for value in values
            root = insert(root, value)
        end
        new(root)
    end
end

function count(node::Nothing)::Integer
    return 0
end

function count(node::BinaryNode)::Integer
    return count(node.left) + count(node.right) + 1
end

function count(tree::BinaryTree)::Integer
    return count(tree.root)
end

function height(node::Nothing)::Integer
    return 0
end

function height(node::BinaryNode)::Integer
    return max(height(node.left), height(node.right)) + 1
end

function height(tree::BinaryTree)::Integer
    return height(tree.root)
end

function find_min(node::BinaryNode)::BinaryNode
    return node.left == nothing ? node : find_min(node.left)
end

function find_max(node::BinaryNode)::BinaryNode
    return node.right == nothing ? node : find_max(node.right)
end

function find(node::Nothing, value)::Nothing
    return nothing
end

function find(node::BinaryNode, value)::BinaryNode
    if value == node.value
        return node
    end
    return find(value < node.value ? node.left : node.right, value)
end

function find(tree::BinaryTree, value)::BinaryNode
    return find(tree.root)
end

function delete(node::Nothing, value)
end

function delete(node::BinaryNode, value)
    if value < node.value
        node.left = delete(node.left, value)
    elseif value > node.value
        node.right = delete(node.right, value)
    else
        if node.left == nothing && node.right == nothing
            return nothing
        elseif node.left == nothing
            return node.right
        elseif node.right == nothing
            return node.left
        else
            child = find_min(node.right)
            node.value = child.value
            node.right = delete(node.right, child.value)
        end
    end
    return node
end

function delete(tree::BinaryTree, value)
    tree.root = delete(tree.root, value)
end

function insert(node::Nothing, value)::BinaryNode
    return BinaryNode(value)
end

function insert(node::BinaryNode, value)::BinaryNode
    if value < node.value
        node.left = insert(node.left, value)
    elseif value > node.value
        node.right = insert(node.right, value)
    end
    return node
end

function insert(tree::BinaryTree, value)
    tree.root = insert(tree.root, value)
end

function inorder(node::Nothing)::Array
    return []
end

function inorder(node::BinaryNode)::Array
    a = inorder(node.left)
    push!(a, node.value)
    append!(a, inorder(node.right))
    return a
end

function inorder(tree::BinaryTree)::Array
    return inorder(tree.root)
end

function leaves(node::Nothing)::Array
    return []
end

function leaves(node::BinaryNode)::Array
    if node.left == nothing && node.right == nothing
        return [node.value]
    end
    a = leaves(node.left)
    append!(a, leaves(node.right))
    return a
end

function leaves(tree::BinaryTree)::Array
    return leaves(tree.root)
end

function paths(node::Nothing)::Array
    return []
end

function paths(node::BinaryNode)::Array
    if node.left == nothing && node.right == nothing
        return [[node.value]]
    end
    a = []
    for path in paths(node.left)
        prepend!(path, node.value)
        append(a, path)
    end
    for path in paths(node.right)
        prepends!(path, node.value)
        append!(a, path)
    end
end

function paths(tree::BinaryTree)::Array
    return paths(tree.root)
end
